src/lib/translators/mongo.ts

```diff
@@ -64,6 +81,155 @@ export function _simplifyAndCondition(filterAndCond: FilterComboAndMongo): Mongo
+          merge = Object.values(lastStep.$project).every(value => Boolean(value) === included);
+        }
+        if (Object.prototype.hasOwnProperty.call(lastStep[stepOperator], key)) {
+          // We do not want to merge two $project with common keys
+          merge = false;
+          break;
+        }
+        if (stepOperator !== '$match') {
+          // We do not want to merge two $project or $addFields with a `step`
+          // key referencing as value a`lastStep` key
+          const valueString: string = JSON.stringify(step[stepOperator][key]);
+          for (const lastKey in lastStep[stepOperator]) {
+            const regex = new RegExp(`.*['"]\\$${lastKey}(\\..+)?['"].*`);
+            if (regex.test(valueString)) {
+              merge = false;
+              break;
+            }
+          }
+        }
+      }
+      if (merge) {
+        // merge $project steps together
+        lastStep[stepOperator] = { ...lastStep[stepOperator], ...step[stepOperator] };
+        continue;
+      }
+    }
+    lastStep = step;
+    outputSteps.push(lastStep);
+    merge = true;
+  }
+  return outputSteps;
+}
+
+function getOperator(op: string) {
+  const operators: PropMap<string> = {
+    '+': '$add',
+    '-': '$subtract',
+    '*': '$multiply',
+    '/': '$divide',
+  };
+  if (operators[op] === undefined) {
+    throw new Error(`Unsupported operator ${op}`);
+  } else {
+    return operators[op];
+  }
+}
+
+/**
+  Transform a formula expression into a MathNode
+  1. Replace in formula all column name between `[]` by a "pseudo"
+  2. Parse the formula into a MathNode
+  3. Replace all pseudo into MathNode by there original name
+*/
+function buildFormulaTree(formula: string): MathNode {
+  const ESCAPE_OPEN = '[';
+  const ESCAPE_CLOSE = ']';
+
+  // 1. Replace in formula all column name between `[]` by a "pseudo"
+  let formulaPseudotised = formula;
+  const pseudo: Record<string, string> = {};
+  let index = 0;
+  const regex = new RegExp(`\\${ESCAPE_OPEN}(.*?)\\${ESCAPE_CLOSE}`, 'g');
+  for (const match of formula.match(regex) || []) {
+    pseudo[`__vqb_col_${index}__`] = match;
+    formulaPseudotised = formulaPseudotised.replace(match, `__vqb_col_${index}__`);
+    index++;
+  }
+
+  // 2. Parse the formula into a MathNode
+  const mathjsTree: MathNode = math.parse(formulaPseudotised);
+
+  // 3. Replace all pseudo into MathNode by there original name
+  mathjsTree.traverse(function(node: MathNode): MathNode {
+    if (node.type === 'SymbolNode') {
+      if (pseudo[node.name]) {
+        node.name = pseudo[node.name].replace(ESCAPE_OPEN, '').replace(ESCAPE_CLOSE, '');
+      }
+    }
+    return node;
+  });
+  return mathjsTree;
+}
+
+function buildConditionTree(
+  cond: S.FilterSimpleCondition | S.FilterComboAnd | S.FilterComboOr,
+): MongoStep {
+  const operatorMapping = {
+    eq: '$eq',
+    ne: '$ne',
+    lt: '$lt',
+    le: '$lte',
+    gt: '$gt',
+    ge: '$gte',
+    in: '$in',
+    nin: '$nin',
```

Comment:
I don't find `$nin` among expression operators: https://docs.mongodb.com/manual/meta/aggregation-quick-reference/#index-of-expression-operators
I think we have no choice than to do a `$not: {$in: ...}}`
This breaks a bit the string map approach, we may have to use small functions instead

Reply:
Oooh thank you for your vigilance!

Reply:
@davinov actually it does exist https://docs.mongodb.com/v4.0/reference/operator/query/nin/index.html
Not sure why it's not referenced on the page you linked... I've tried it with a UAT it's all good :)